File: app/src/main/java/br/pedroso/upcomingmovies/network/di/NetworkModule.java
package br.pedroso.upcomingmovies.network.di;

import android.app.Application;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import javax.inject.Singleton;

import br.pedroso.upcomingmovies.network.deserializer.UpcomingMoviesResultDeserializer;
import br.pedroso.upcomingmovies.network.interceptors.ApiKeyInterceptor;
import br.pedroso.upcomingmovies.network.services.TheMovieDbService;
import dagger.Module;
import dagger.Provides;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava3.RxJava3CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class NetworkModule {
    private final String apiKey;

    private final long CACHE_SIZE = 10 * 1024 * 1024;

    public NetworkModule(String apiKey) {
        this.apiKey = apiKey;
    }

    @Provides
    @Singleton
    Cache provideHttpCache(Application application) {
        Cache cache = new Cache(application.getCacheDir(), CACHE_SIZE);
        return cache;
    }

    @Provides
    @Singleton
    ApiKeyInterceptor provideApiKeyInterceptor() {
        return new ApiKeyInterceptor(apiKey);
    }

    @Provides
    @Singleton
    OkHttpClient provideOkHttpClient(Cache cache, ApiKeyInterceptor apiKeyInterceptor) {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();

        builder.cache(cache);

        builder.interceptors().add(apiKeyInterceptor);

        return builder.build();
    }

    @Provides
    @Singleton
    Gson provideGson() {
        GsonBuilder gsonBuilder = new GsonBuilder();

        gsonBuilder.registerTypeAdapter(UpcomingMoviesResultDeserializer.getRetrofitMovieEntitiesListType(), new UpcomingMoviesResultDeserializer());

        return gsonBuilder.create();
    }

    @Provides
    @Singleton
    Retrofit provideRetrofit(Gson gson, OkHttpClient httpClient) {
        Retrofit retrofit = new Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava3CallAdapterFactory.create())
                .baseUrl(TheMovieDbService.BASE_URL)
                .client(httpClient)
                .build();

        return retrofit;
    }

    @Provides
    @Singleton
    TheMovieDbService provideTheMovieDbService(Retrofit retrofit) {
        return retrofit.create(TheMovieDbService.class);
    }
}

File: app/src/main/kotlin/br/pedroso/upcomingmovies/network/di/NetworkModule.kt
package br.pedroso.upcomingmovies.network.di

import android.app.Application
import br.pedroso.upcomingmovies.BuildConfig
import br.pedroso.upcomingmovies.network.interceptors.ApiKeyInterceptor
import br.pedroso.upcomingmovies.network.services.TheMovieDbService
import br.pedroso.upcomingmovies.network.services.TheMovieDbServiceKtor
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import io.ktor.client.HttpClient
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.defaultRequest
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json
import okhttp3.Cache
import javax.inject.Singleton

@InstallIn(SingletonComponent::class)
@Module
class NetworkModule {
    @Provides
    @Singleton
    fun provideHttpCache(application: Application): Cache {
        val cache = Cache(application.cacheDir, CACHE_SIZE)
        return cache
    }

    @Provides
    @Singleton
    fun provideApiKeyInterceptor(): ApiKeyInterceptor {
        return ApiKeyInterceptor(BuildConfig.MOVIES_DB_API_KEY)
    }

    @Provides
    @Singleton
    fun provideJson(): Json {
        return Json {
            ignoreUnknownKeys = true
            coerceInputValues = true
        }
    }

    @Provides
    @Singleton
    fun provideHttpClient(
        cache: Cache?,
        apiKeyInterceptor: ApiKeyInterceptor,
        json: Json,
    ): HttpClient {
        return HttpClient(OkHttp) {
            expectSuccess = true

            engine {
                addInterceptor(apiKeyInterceptor)
                config {
                    cache(cache)
                }
            }

            defaultRequest {
                url(TheMovieDbService.BASE_URL)
            }

            install(ContentNegotiation) {
                json(json)
            }
        }
    }

    @Provides
    @Singleton
    fun provideTheMovieDbService(httpClient: HttpClient): TheMovieDbService {
        return TheMovieDbServiceKtor(httpClient)
    }

    companion object {
        private const val CACHE_SIZE = (10 * 1024 * 1024).toLong()
    }
}
